packages/lib/gpcircuits/src/multituple.ts

```diff
@@ -0,0 +1,232 @@
+    // `paramTupleArity` elements or less,we require one tuple.
+    // Else, we require one tuple for the first `paramTupleArity` elements
+    // plus one for every additional `paramTupleArity` - 1 elements, since
+    // the first tuple element for every subsequent tuple will be reserved
+    // for the index pointing to the preceding tuple.
+    return Math.ceil((tupleArity - 1) / (paramTupleArity - 1));
+  }
+}
+
+/**
+ * Determines the maximum tuple arity (of a single tuple) representable
+ * by the given parameters.
+ * Throws a `RangeError` is the parameters are not proper.
+ */
+export function maxTupleArity(
+  paramMaxTuples: number,
+  paramTupleArity: number
+): number {
+  if (paramTupleArity < 2) {
+    throw new RangeError("The tuple arity parameter must be at least 2.");
+  } else if (paramMaxTuples < 0) {
+    throw new RangeError(
+      "The maximum tuple number parameter must be nonnegative."
+    );
+  } else {
+    return paramMaxTuples === 0
+      ? // If we have no tuples at our disposal, we cannot represent any tuples,
+        // thus return arity 0.
+        0
+      : // Else we have `paramTupleArity` slots in the first tuple at our
+        // disposal and `paramTupleArity` - 1 slots in every subsequent tuple,
+        // since the first slot of these tuples will contain the index pointing
+        // to the preceding tuple.
+        paramTupleArity + (paramTupleArity - 1) * (paramMaxTuples - 1);
+  }
+}
+
+/**
+ * This procedure takes an N-tuple of `indices` (referring
+ * to elements of some array of values) and returns its
+ * representation as an appropriately linked sequence of
+ * tuples of arity `paramMaxArity`, where these tuples
+ * are indexed starting at `firstAvailableTupleIndex` and the
+ * first index is used as padding. 'Appropriately linked'
+ * means that the first slot of every output tuple after
+ * the first has an index referring to the preceding
+ * tuple.
+ *
+ * Examples:
+ * computeTupleIndices(2, 5, [1, 3, 4]) === [[1, 3], [5, 4]]
+ * computeTupleIndices(3, 5, [0, 1, 4, 2]) === [[0, 1, 4], [5, 2, 0]]
+ * computeTupleIndices(4, 6, [3, 4, 2, 1, 5]) === [[3, 4, 2, 1], [6, 5, 3, 3]]
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param firstAvailableTupleIndex the index of the first output tuple in the combined array of
+ * entry and tuple value hashes
+ * @param indices the entry value indices forming the input tuple
+ * @returns list of index tuples of arity `paramTupleArity` representing the input tuple
+ * @throws RangeError if the parameters are not in the proper ranges.
+ */
+export function computeTupleIndices(
+  paramTupleArity: number,
+  firstAvailableTupleIndex: number,
+  indices: number[]
+): number[][] {
+  if (indices.some((i) => i >= firstAvailableTupleIndex)) {
+    throw new RangeError(
+      "The first available tuple index must be strictly greater than the given indices."
+    );
+  }
+
+  // Note the first index.
```

Comment:
Why?  I assume for padding later.
If it doesn't say why, then this comment doesn't add anything which the code doesn't already say.